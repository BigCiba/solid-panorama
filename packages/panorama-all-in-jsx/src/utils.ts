import { NodePath } from '@babel/core';
import { addNamed } from '@babel/helper-module-imports';
import * as t from '@babel/types';

export function registerImportMethod(
    path: NodePath,
    name: string,
    moduleName: string
) {
    let imports = path.scope.getProgramParent().getData('imports');
    if (!imports) {
        imports = new Map();
        path.scope.getProgramParent().setData('imports', imports);
    }
    if (!imports.has(`${moduleName}:${name}`)) {
        let id = addNamed(path, name, moduleName, {
            nameHint: `_$${name}`
        });
        imports.set(`${moduleName}:${name}`, id);
        return id;
    } else {
        let iden = imports.get(`${moduleName}:${name}`);
        return t.cloneNode(iden);
    }
}
