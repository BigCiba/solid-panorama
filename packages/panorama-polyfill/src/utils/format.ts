// Based on https://github.com/browserify/node-util/blob/4b1c0c79790d9968eabecd2e9c786454713e200f/util.js#L33
export function format(value?: unknown, ...substitutions: unknown[]) {
    if (typeof value !== 'string') {
        return [value, ...substitutions].map(v => inspect(v)).join(' ');
    }

    let count = 0;
    let result = String(value).replace(/%[sdj%]/g, x => {
        count++;
        if (x === '%%') return '%';
        if (substitutions.length === 0) return x;

        switch (x) {
            case '%s':
                return String(substitutions.shift());

            case '%d':
                return String(Number(substitutions.shift()));

            case '%j':
                try {
                    return JSON.stringify(substitutions.shift());
                } catch {
                    return '[Circular]';
                }

            default:
                return x;
        }
    });

    for (const x of substitutions) {
        if (typeof x !== 'object' || x === null) {
            result += ` ${x}`;
        } else {
            result += ` ${inspect(x)}`;
        }
    }

    return result;
}

export function formatx(value?: unknown, ...substitutions: unknown[]) {
    if (typeof value !== 'string') {
        return [value, ...substitutions].map(v => inspect(v, true)).join(' ');
    }

    let result = String(value).replace(/%[sdj%]/g, x => {
        if (x === '%%') return '%';
        if (substitutions.length === 0) return x;

        switch (x) {
            case '%s':
                return String(substitutions.unshift());

            case '%d':
                return String(Number(substitutions.unshift()));

            case '%j':
                try {
                    return JSON.stringify(substitutions.unshift());
                } catch {
                    return '[Circular]';
                }

            default:
                return x;
        }
    });

    for (const x of substitutions) {
        if (typeof x !== 'object' || x === null) {
            result += ` ${x}`;
        } else {
            result += ` ${inspect(x)}`;
        }
    }

    return result;
}

function inspect(value: unknown, expand?: boolean, tab = '') {
    let result = '';
    if (typeof value === 'string') {
        result = `"${value}"`;
    } else if (typeof value === 'number' || typeof value === 'boolean') {
        result = `${value}`;
    } else if (typeof value === 'function') {
        result = isClass(value)
            ? `[class ${value.name}]`
            : `[function ${value.name}]`;
    } else if (typeof value === 'symbol') {
        result = value.toString();
    } else if (typeof value === 'undefined') {
        result = `undefined`;
    } else if (typeof value === 'bigint') {
        result = `[bigint ${value.toString()}]`;
    } else if (typeof value === 'object') {
        if (value === null) {
            result = 'null';
        } else if (Array.isArray(value)) {
            let list = [];
            for (const v of value) {
                list.push(
                    tab + inspect(v, expand, expand ? tab + '    ' : tab)
                );
            }
            if (expand) {
                result += '[\n';
                result += list.map(v => '    ' + v).join(',\n');
                result += '\n' + tab + ']';
            } else {
                result = `[ ${list.join(', ')} ]`;
            }
        } else {
            let list: string[] = [];
            let prefix = '';
            // Map
            if (value instanceof Map) {
                prefix = '[Map]';
                for (const [k, v] of value.entries()) {
                    let key = '';
                    if (typeof k === 'object') {
                        key = Array.isArray(k) ? '[Array]' : '[Object]';
                    } else if (k.toString) {
                        key = k.toString();
                    }
                    list.push(
                        `${tab}${key}: ${inspect(
                            v,
                            expand,
                            expand ? tab + '    ' : tab
                        )}`
                    );
                }
            } else if (value instanceof Set) {
                prefix = '[Set]';
                for (const v of value.values()) {
                    list.push(
                        `${tab}${inspect(
                            v,
                            expand,
                            expand ? tab + '    ' : tab
                        )}`
                    );
                }
            } else {
                const isPanel = isPanelBase(value);
                for (const [k, v] of Object.entries(value)) {
                    if (isPanel) {
                        console.log(value, k, v);
                    }
                    if (k === 'style' && isPanel) {
                        list.push(`${tab}${k}: [VCSSStyleDeclaration]`);
                        continue;
                    }
                    list.push(
                        `${tab}${k}: ${inspect(
                            v,
                            expand,
                            expand ? tab + '    ' : tab
                        )}`
                    );
                }
            }

            if (expand) {
                result += prefix + '{\n';
                result += list.map(v => '    ' + v).join(',\n');
                result += '\n' + tab + '}';
            } else {
                result = prefix + `{ ${list.join(', ')} }`;
            }
        }
    }
    return result;
}

const isPanelBase = (value: object): value is PanelBase =>
    'paneltype' in value &&
    'rememberchildfocus' in value &&
    'SetPanelEvent' in value;

function isClass(value: unknown) {
    if (typeof value !== 'function') return false;
    const descriptor = Object.getOwnPropertyDescriptor(value, 'prototype');
    if (!descriptor) return false;
    return !descriptor.writable;
}
