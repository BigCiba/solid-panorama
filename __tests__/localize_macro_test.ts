import { parseMacros } from './utils';
import { describe, expect, test } from '@jest/globals';
import {
    getLocalizationTable,
    autoApplyToLocalizationFile
} from '../packages/panorama-all-in-jsx/src/localize.macro';
import { join } from 'path';
import { mkdir, readFile } from 'fs/promises';
import { KeyValues } from 'easy-keyvalues';
import { existsSync } from 'fs';

describe('localize_macro', function () {
    test('getLocalizationTable', async function () {
        const result = parseMacros(
            `
            import localize from '../packages/panorama-all-in-jsx/src/localize.macro';
            const token_a = localize('test_token_a', 'this is a', '这是a', 'Это a')
            const token_b = localize('#test_token_b', 'this is b', '这是b', 'Это b')
            const token_c = localize('#', 'this is c', '这是c', 'Это c')
            const token_d = localize('', 'this is d', '这是d', 'Это d')
        `,
            __filename,
            {
                'panorama-all-in-jsx': {
                    localize_language_argv: ['english', 'schinese', 'russian']
                }
            }
        );
        expect(result).toMatchSnapshot();

        expect(getLocalizationTable()).toStrictEqual({
            test_token_a: {
                english: 'this is a',
                schinese: '这是a',
                russian: 'Это a'
            },
            test_token_b: {
                english: 'this is b',
                schinese: '这是b',
                russian: 'Это b'
            },
            token_f780dc967d9b71be: {
                english: 'this is c',
                schinese: '这是c',
                russian: 'Это c'
            },
            token_fa636963348fd7d0: {
                english: 'this is d',
                schinese: '这是d',
                russian: 'Это d'
            }
        });

        const basePath = join(__dirname, 'localization');
        await mkdir(basePath, { recursive: true });
        const kv = KeyValues.CreateRoot();
        kv.CreateChild('lang', []);
        kv.Save(join(basePath, 'addon_schinese.txt'), 'utf8');
        kv.FindKey('lang')?.CreateChild('addon_game_name', 'My Game');
        kv.Save(join(basePath, 'addon_english.txt'), 'utf8');
        await autoApplyToLocalizationFile(basePath);
        expect(await readFile(join(basePath, 'addon_english.txt'), 'utf8'))
            .toBe(`"lang"
{
    "addon_game_name"           "My Game"
    "test_token_a"              "this is a"
    "test_token_b"              "this is b"
    "token_f780dc967d9b71be"    "this is c"
    "token_fa636963348fd7d0"    "this is d"
}`);
        expect(await readFile(join(basePath, 'addon_schinese.txt'), 'utf8'))
            .toBe(`"lang"
{
    "test_token_a"              "这是a"
    "test_token_b"              "这是b"
    "token_f780dc967d9b71be"    "这是c"
    "token_fa636963348fd7d0"    "这是d"
}`);
        expect(await readFile(join(basePath, 'addon_russian.txt'), 'utf8'))
            .toBe(`"lang"
{
    "test_token_a"              "Это a"
    "test_token_b"              "Это b"
    "token_f780dc967d9b71be"    "Это c"
    "token_fa636963348fd7d0"    "Это d"
}`);
    });
});
