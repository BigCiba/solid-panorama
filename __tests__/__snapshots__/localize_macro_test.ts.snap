// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`localize_macro getLocalizationList 1`] = `
""use strict";

const token_a = "#test_token_a";
const token_b = "#test_token_b";
const token_c = "#token_f780dc967d9b71be";
const token_d = "#token_fa636963348fd7d0";"
`;
